File: 0001-0500/0023/0023.merge-k-sorted-lists.cs
/*
 * @lc app=leetcode id=23 lang=csharp
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeKLists(ListNode[] lists) {
        if(lists.Length == 0) return null;
        if(lists.Length == 1) return lists[0];
        if(lists.Length == 2) return MergeTwoLists(lists[0], lists[1]);
        int mid = lists.Length / 2;
        ListNode[] left = new ListNode[mid];
        ListNode[] right = new ListNode[lists.Length - mid];
        Array.Copy(lists, 0, left, 0, mid);
        Array.Copy(lists, mid, right, 0, lists.Length - mid);
        return MergeTwoLists(MergeKLists(left), MergeKLists(right));
    }

    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {
        if(list1 == null) return list2;
        if(list2 == null) return list1;
        if(list1.val < list2.val) {
            list1.next = MergeTwoLists(list1.next, list2);
            return list1;
        } else {
            list2.next = MergeTwoLists(list1, list2.next);
            return list2;
        }
    }
}
// @lc code=end

File: 0001-0500/0023/0023.merge-k-sorted-lists.js
/*
 * @lc app=leetcode id=23 lang=javascript
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
    if (lists.length === 0) return null;
    else if (lists.length === 1) return lists[0];

    let left, right;
    if (lists.length > 2) {
        const mid = Math.floor(lists.length / 2);
        left = mergeKLists(lists.slice(0, mid));
        right = mergeKLists(lists.slice(mid));
    } else {
        left = lists[0];
        right = lists[1];
    }

    if (left === null) return right;
    if (right === null) return left;

    let head = null;
    let curr = null;

    while (left !== null && right !== null) {
        if (left.val <= right.val) {
            if (head === null) {
                head = left;
                curr = head;
                left = left.next;
            } else {
                curr.next = left;
                left = left.next;
                curr = curr.next;
            }
        } else {
            if (head === null) {
                head = right;
                curr = head;
                right = right.next;
            } else {
                curr.next = right;
                right = right.next;
                curr = curr.next;
            }
        }
    }

    if (left !== null) {
        curr.next = left;
    }

    if (right !== null) {
        curr.next = right;
    }

    return head;
};
// @lc code=end

